src/ram/src/ram.cpp

```diff
@@ -546,6 +550,44 @@ void RamGen::generate(const int bytes_per_word,
   ram_grid.setOrigin(ram_origin);
   ram_grid.gridInit();
 
+  if (tapcell_) {
+    if (ram_grid.getRowWidth() <= max_tap_dist) {
+      auto tapcell_layout = std::make_unique<Layout>(odb::vertical);
+      for (int i = 0; i < word_count; ++i) {
+        auto tapcell_cell = std::make_unique<Cell>();
+        makeCellInst(tapcell_cell.get(),
+                     "tapcell",
+                     fmt::format("cell_{}", i),
+                     tapcell_,
+                     {});
+        tapcell_layout->addCell(std::move(tapcell_cell));
+      }
+      ram_grid.insertLayout(std::move(tapcell_layout), 0); 
+    } else {
+      int nearest_tap = 0;
+      for (int col = 0; col < ram_grid.numLayouts(); ++col) {
+        if (nearest_tap >= max_tap_dist) {
+          auto tapcell_layout = std::make_unique<Layout>(odb::vertical);
+          for (int i = 0; i < word_count; ++i) {
+            auto tapcell_cell = std::make_unique<Cell>();
+            makeCellInst(tapcell_cell.get(),
+                         "tapcell",
+                         fmt::format("cell{}_{}", col, i),
+                         tapcell_,
+                         {});
+            tapcell_layout->addCell(std::move(tapcell_cell));
```

Comment:
Maybe you should make a function like `generateTapColumn()` which generates the tapcell column, then you can insert it.